hypothesis-python/tests/cover/test_deferred_strategies.py

```diff
@@ -42,9 +42,15 @@ def test_mutual_recursion():
 
 
 def test_errors_on_non_function_define():
-    x = st.deferred(1)
+    # if argument is not a function we should raise an error when building the strategy.
     with pytest.raises(InvalidArgument):
-        x.example()
+        st.deferred(1)
```

Comment:
Ah... this is why the validation happens in the `wrapped_strategy` property: so that the error is raised when the test is executed, rather than at test collection time.

We actually *shouldn't* change that, but it would be really useful to add a comment to both the `deferred` function (in `strategies/_internal/core.py`) and above the validation logic in `wrapped_strategy`.

I'd also like to add a *deprecation* (with `_settings.note_deprecation`) if the function has arguments - I'll turn that into an error in Hypothesis 5.  `inspect.signature` doesn't exist yet, but you can use `hypothesis.internal.compat.getfullargspec` as `sig = getfullargspec(f); if sig.args or sig.varargs or sig.varkw or sig.kwonlyargs: note_deprecation(..., since="RELEASEDAY")`.  Admittedly this is pretty clumsy, but if that's what it takes for compatibility...

Reply:
> Ah... this is why the validation happens in the `wrapped_strategy` property: so that the error is raised when the test is executed, rather than at test collection time.

wait we want that? 
Ok yeah wow i can't believe i didn't notice that before... invalid strategies come back as failures.

I'd made the assumption we wanted invalid args to strategies to be caught as errors on collection not test failures ( since the test cannot fail if it has not run).

I'm sure there's a good reason we want them to come back as failures... maybe there's an issue or something in the history explaining it. I'll have a look ...

Reply:
> I'm sure there's a good reason we want them to come back as failures... maybe there's an issue or something in the history explaining it.

We don't want to have collection errors when you were trying to run a different test.

It's basically an arbitrary choice, but we've found this way give significantly better integration and error output from various test runners.

Reply:
---
> We actually shouldn't change that, but it would be really useful to add a comment to both the deferred function (in strategies/_internal/core.py) and above the validation logic in wrapped_strategy.

is there a point in validating zero-args in `wrapped_strategy` if it's actually called in there?  If we want better error messaging we can just `try:except:raise` around the call.  especially with all the stuff below...

>I'd also like to add a deprecation (with _settings.note_deprecation) if the function has arguments - I'll turn that into an error in Hypothesis 5. inspect.signature doesn't exist yet, but you can use hypothesis.internal.compat.getfullargspec as sig = getfullargspec(f); if sig.args or sig.varargs or sig.varkw or sig.kwonlyargs: note_deprecation(..., since="RELEASEDAY"). Admittedly this is pretty clumsy, but if that's what it takes for compatibility...

`inspect.getfullargspec` (which is i see what `hypothesis.internal.compat.getfullargspec` based on) doesn't handle default args well or optargs  as well as binding to signature... which is kinda annoying. but if we see a reason to do the check at all and add it in before jan1 i can parse the opt types to filter defaults etc.

I wouldn't think we would not wanna disallow kwargs with defaults or optargs  if we don't have to:
silly example...

```python
from hypothesis import strategies as st


def words(prefix='', suffix=''):
    wrds = ['set', 'pair', 'mind']
    if prefix == 're': 
        wrds.append('pent')
    return st.sampled_from(wrds).map(lambda word: f'{prefix}{word}{suffix}')

a = st.deferred(words).example()
b = st.deferred(lambda : words(prefix='re')).example()
c = st.deferred(lambda : words(prefix='re', suffix='s')).example()
```

also while writing above silly example i notice the `inspect.isfunction` we use does not allow partials or bound methods... 
is that a thing?

Reply:
> It's basically an arbitrary choice, but we've found this way give significantly better integration and error output from various test runners.

yeah that's fair enough. I basically only use pytest so i haven't seen how they are handled elsewhere... but hypothesis is bigger than just a pytest plugin.

Reply:
Hmm... I think the easiest solution is probably just to say "it's good enough as it is" and close the PR and the issue :confused: 

Sorry for taking so long to work that out.

Reply:
If none of this is required at all i'm happy to drop the pr... no love lost and i learned something valuable about how and when we raise exceptions.

Reply:
> Hmm... I think the easiest solution is probably just to say "it's good enough as it is" and close the PR and the issue 😕
> 
> Sorry for taking so long to work that out.

as above... no worries at all.